pkg/ccl/changefeedccl/changefeed_test.go

```diff
@@ -966,6 +966,159 @@ func TestChangefeedSchemaChangeNoBackfill(t *testing.T) {
+
+		// Ensure events are consumed for sinks that don't buffer (ex: Kafka)
+		g := ctxgroup.WithContext(context.Background())
+		g.Go(func() error {
+			for {
+				_, err := baseFeed.Next()
+				if err != nil {
+					return err
+				}
+			}
+		})
+		defer func() {
+			closeFeed(t, baseFeed)
+			_ = g.Wait()
+		}()
+
+		// Helper to read job progress
+		loadProgress := func() jobspb.Progress {
+			jobID := jobFeed.JobID()
+			job, err := jobRegistry.LoadJob(context.Background(), jobID)
+			require.NoError(t, err)
+			return job.Progress()
+		}
+
+		// Wait for backfill to complete
+		testutils.SucceedsSoon(t, func() error {
+			prog := loadProgress()
+			if p := prog.GetHighWater(); p != nil && !p.IsEmpty() {
+				return nil
+			}
+			return errors.New("waiting for highwater")
+		})
+
+		// Should have grabbed a span within the table
+		require.True(t, middleSpan.Key.Compare(tableSpan.Key) == 1)
+		require.True(t, middleSpan.EndKey.Compare(tableSpan.EndKey) == -1)
+
+		// Pause job and setup overrides to force a checkpoint
+		require.NoError(t, jobFeed.Pause())
+
+		// Checkpoint progress frequently, allow a large enough checkpoint, and
+		// reduce the lag threshold to allow lag checkpointing to trigger
+		changefeedbase.FrontierCheckpointFrequency.Override(
+			context.Background(), &f.Server().ClusterSettings().SV, 10*time.Millisecond)
+		changefeedbase.FrontierCheckpointMaxBytes.Override(
+			context.Background(), &f.Server().ClusterSettings().SV, 100<<20)
+		changefeedbase.FrontierHighwaterLagCheckpointThreshold.Override(
+			context.Background(), &f.Server().ClusterSettings().SV, 10*time.Millisecond)
+
+		knobs.ShouldSkipResolved = func(r *jobspb.ResolvedSpan) bool {
+			// Only resolve spans around the middle "lagging" span
+			if rnd.Intn(10) < 5 && r.Span.Key.Compare(middleSpan.Key) <= 0 {
+				r.Span.EndKey = middleSpan.Key
+			} else if r.Span.EndKey.Compare(middleSpan.EndKey) >= 1 {
+				r.Span.Key = middleSpan.EndKey
+			} else {
+				return true
+			}
+			return false
+		}
+
+		require.NoError(t, jobFeed.Resume())
+
+		// Should eventually checkpoint
+		testutils.SucceedsSoon(t, func() error {
+			progress := loadProgress()
+			if p := progress.GetChangefeed(); p != nil && p.Checkpoint != nil && len(p.Checkpoint.Spans) >= 2 {
+				return nil
+			}
+			return errors.New("waiting for checkpoint around missing span")
+		})
+
+		// Pause the job and read and verify the latest checkpoint information.
+		require.NoError(t, jobFeed.Pause())
+		progress := loadProgress()
+		jobCheckpoint := progress.GetChangefeed().Checkpoint
+		require.True(t, progress.GetHighWater().Less(jobCheckpoint.Timestamp))
+		var checkpointSpanGroup roachpb.SpanGroup
+		checkpointSpanGroup.Add(jobCheckpoint.Spans...)
+
+		// Allow the changefeed to catch up without re-emitting checkpointed spans
+		knobs.ShouldSkipResolved = func(r *jobspb.ResolvedSpan) bool {
+			require.False(t, checkpointSpanGroup.Encloses(r.Span) && r.Timestamp.LessEq(jobCheckpoint.Timestamp))
```

Comment:
This doesn't end up verifying anything because we just immediately get a ResolvedSpan for the whole table at a timestamp thats ahead of even the checkpoint.  I'm not sure how to force the case where we get an intermediate ResolvedSpan of just the lagging area.

Reply:
Add a ShouldSkipResolved override that skips everything but the middle span. That should complete the frontier.

Reply:
Right now we just end up getting one ResolvedSpan thats the whole table at a timestamp thats later than the checkpoint, rather than anything specific to the middle span.